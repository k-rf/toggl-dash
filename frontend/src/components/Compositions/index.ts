export * from "./Form";
export * from "./RhfTextField";
