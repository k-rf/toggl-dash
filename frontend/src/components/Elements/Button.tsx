import { Button as MuiButton, ButtonProps as MuiButtonProps } from "@mui/material";

type Props = Pick<
  MuiButtonProps,
  "children" | "color" | "disabled" | "fullWidth" | "onClick" | "variant"
>;

export const Button = (props: Props) => {
  return <MuiButton {...props} />;
};
