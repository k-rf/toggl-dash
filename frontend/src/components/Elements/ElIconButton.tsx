import { IconButton, IconButtonProps } from "@mui/material";

type Props = Pick<
  IconButtonProps,
  "children" | "onClick" | "onMouseDown" | "onMouseUp" | "onMouseLeave" | "disableTouchRipple"
>;

export const ElIconButton = (props: Props) => {
  return <IconButton {...props} />;
};

export type { Props as ElIconButtonProps };
