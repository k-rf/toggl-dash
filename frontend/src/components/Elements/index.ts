export * from "./ElButton";
export * from "./ElPaper";
export * from "./ElTextField";
export * from "./ElAccordion";
