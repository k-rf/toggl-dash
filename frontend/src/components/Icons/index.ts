export * from "./ExpandMoreIcon";
export * from "./PlayArrowIcon";
export * from "./VisibilityIcon";
export * from "./VisibilityOffIcon";
