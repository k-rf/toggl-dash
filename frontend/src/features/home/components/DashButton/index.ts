export * from "./DashButton";
export * from "./DashButtonDetailListItem";
