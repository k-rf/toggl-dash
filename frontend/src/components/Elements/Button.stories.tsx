import { Box, Stack } from "@mui/material";
import type { ComponentMeta, ComponentStoryObj } from "@storybook/react";

import { Button } from "./Button";

type Meta = ComponentMeta<typeof Button>;
type Story = ComponentStoryObj<typeof Button>;

export default { title: "Elements/Button", component: Button } as Meta;

export const Default: Story = {
  args: {
    children: "ボタン",
    color: "primary",
  },
};

export const Variant: Story = {
  ...Default,
  args: {
    ...Default.args,
  },
  render: (args) => {
    return <Buttons {...args} />;
  },
};

export const Disabled: Story = {
  ...Default,
  args: {
    ...Default.args,
    disabled: true,
  },
  render: (args) => {
    return <Buttons {...args} />;
  },
};

export const Colors: Story = {
  ...Default,
  args: {
    ...Default.args,
    fullWidth: true,
  },
  render: (args) => {
    return (
      <Stack direction={"column"} spacing={2}>
        <Buttons {...args} color="primary" width={120}>
          primary
        </Buttons>
        <Buttons {...args} color="secondary" width={120}>
          secondary
        </Buttons>
        <Buttons {...args} color="success" width={120}>
          success
        </Buttons>
        <Buttons {...args} color="warning" width={120}>
          warning
        </Buttons>
        <Buttons {...args} color="error" width={120}>
          error
        </Buttons>
        <Buttons {...args} color="info" width={120}>
          info
        </Buttons>
      </Stack>
    );
  },
};

const Buttons = (args: Story["args"] & { width?: number }) => {
  const { width, ...props } = args;
  return (
    <Stack direction={"row"} spacing={2}>
      <Box width={width}>
        <Button {...props} variant="contained" />
      </Box>
      <Box width={width}>
        <Button {...props} variant="outlined" />
      </Box>
      <Box width={width}>
        <Button {...props} variant="text" />
      </Box>
    </Stack>
  );
};
