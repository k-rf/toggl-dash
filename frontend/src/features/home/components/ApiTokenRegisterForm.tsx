import { InputAdornment, Stack } from "@mui/material";
import { KeyboardEventHandler, MouseEventHandler, useState } from "react";

import { Form, HandleSubmit, Methods } from "~/components/Compositions";
import { ElButton, ElTextField } from "~/components/Elements";
import { ElIconButton } from "~/components/Elements/ElIconButton";
import { ElTypography } from "~/components/Elements/ElTypography";
import { VisibilityIcon, VisibilityOffIcon } from "~/components/Icons";
import { z } from "~/lib/zod";

const schema = z.object({
  apiToken: z.string().min(1),
});

type ApiTokenValues = z.infer<typeof schema>;

type Props = { onSubmit?: (value: ApiTokenValues) => Promise<void> | void };

export const ApiTokenRegisterForm = (props: Props) => {
  const [visible, setVisible] = useState(false);

  const submit = (methods: Methods<ApiTokenValues>) => {
    return methods.handleSubmit(async (value) => {
      methods.reset({ apiToken: "" });

      await props.onSubmit?.(value);
    });
  };

  const handleSubmit: HandleSubmit<ApiTokenValues, MouseEventHandler> = (methods) => {
    return submit(methods);
  };

  const handleKeyDown: HandleSubmit<ApiTokenValues, KeyboardEventHandler> = (methods) => {
    return (e) => {
      if (e.code === "Enter") {
        e.preventDefault();

        if (e.ctrlKey) {
          return submit(methods)(e);
        }
      }
    };
  };

  const handleShow = () => {
    setVisible(true);
  };

  const handleMask = () => {
    setVisible(false);
  };

  return (
    <Form<ApiTokenValues> options={{ mode: "all" }} schema={schema}>
      {(methods) => {
        const { error } = methods.getFieldState("apiToken");
        const {
          errors: { apiToken },
          isValid,
        } = methods.formState;

        return (
          <Stack spacing={1}>
            <ElTypography>Toggl の API トークンを入力してください</ElTypography>
            <ElTextField
              fullWidth
              placeholder="API トークン"
              type={visible ? "text" : "password"}
              error={Boolean(error)}
              onKeyDown={handleKeyDown(methods)}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <ElIconButton
                      disableTouchRipple
                      onMouseDown={handleShow}
                      onMouseUp={handleMask}
                      onMouseLeave={handleMask}
                      data-testid="api-token-register-form-visibility-toggle-icon-button"
                    >
                      {visible ? <VisibilityOffIcon /> : <VisibilityIcon />}
                    </ElIconButton>
                  </InputAdornment>
                ),
              }}
              registration={methods.register("apiToken")}
            />
            <ElButton
              disabled={!isValid}
              onClick={handleSubmit(methods)}
              data-testid="api-token-register-form-submit-button"
            >
              保存する
            </ElButton>
            <ElTypography variant="body2" color="error">
              {apiToken?.message}
            </ElTypography>
          </Stack>
        );
      }}
    </Form>
  );
};

export type { Props as ApiTokenRegisterFormProps };
