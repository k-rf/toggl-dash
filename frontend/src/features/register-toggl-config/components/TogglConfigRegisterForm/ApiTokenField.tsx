import { InputAdornment } from "@mui/material";
import { KeyboardEventHandler, useState } from "react";
import { UseFormRegisterReturn, FieldError } from "react-hook-form";

import { ElTextField } from "~/components/Elements";
import { ElIconButton } from "~/components/Elements/ElIconButton";
import { VisibilityOffIcon, VisibilityIcon } from "~/components/Icons";

type Props = {
  error?: FieldError;
  onKeyDown: KeyboardEventHandler<Element>;
  registration: Partial<UseFormRegisterReturn>;
};

export const ApiTokenField = (props: Props) => {
  const [visible, setVisible] = useState(false);

  const handleShow = () => {
    setVisible(true);
  };

  const handleMask = () => {
    setVisible(false);
  };

  return (
    <ElTextField
      id="api-token-field"
      fullWidth
      size="small"
      placeholder="API トークン"
      type={visible ? "text" : "password"}
      error={Boolean(props.error)}
      onKeyDown={props.onKeyDown}
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <ElIconButton
              disableTouchRipple
              onMouseDown={handleShow}
              onMouseUp={handleMask}
              onMouseLeave={handleMask}
              data-testid="api-token-register-form-visibility-toggle-icon-button"
            >
              {visible ? <VisibilityOffIcon /> : <VisibilityIcon />}
            </ElIconButton>
          </InputAdornment>
        ),
      }}
      registration={props.registration}
    />
  );
};
