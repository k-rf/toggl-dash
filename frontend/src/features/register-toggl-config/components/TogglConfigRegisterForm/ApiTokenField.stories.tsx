import type { Meta, StoryObj } from "@storybook/react";

import { ApiTokenField } from "./ApiTokenField";

export default { component: ApiTokenField } as Meta<typeof ApiTokenField>;

type Story = StoryObj<typeof ApiTokenField>;

export const Default: Story = {};
