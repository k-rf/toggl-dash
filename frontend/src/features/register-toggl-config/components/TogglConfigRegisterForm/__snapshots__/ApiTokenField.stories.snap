// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/Register-toggl-config/Components/TogglConfigRegisterForm/ApiTokenField > Default 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id=":r2:"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="API トークン"
      required={false}
      type="password"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="api-token-register-form-visibility-toggle-icon-button"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="VisibilityIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-ihdtdm"
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
